frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -3956,6 +4011,145 @@ bool isTypeDefaultInitializable(Context* context, const Type* t) {
   return isTypeDefaultInitializableQuery(context, t);
 }
 
+void getCopyOrAssignableInfo(Context* context, const Type* t,
+                                    bool* fromConst, bool* fromRef,
+                                    bool checkCopyable);
+
+// checkCopyable is true for copyable, false for assignable.
+// Return: First is from const, second is from ref.
+static const std::pair<bool, bool>& getCopyOrAssignableInfoQuery(
+    Context* context, const Type* t, bool checkCopyable) {
+  QUERY_BEGIN(getCopyOrAssignableInfoQuery, context, t, checkCopyable);
+
+  CHPL_ASSERT((t->isClassType() || t->isCompositeType()) &&
+              "should only be getting called for class or composite types");
+
+  // Inspect type for either kind of copyability/assignability.
+  bool fromConst = false;
+  bool fromRef = false;
+  if (auto classTy = t->toClassType()) {
+    if (classTy->decorator().isManaged() &&
+        classTy->manager()->isAnyOwnedType()) {
+      // Owned class types are copyable from ref iff they are nilable.
```

Comment:
Wherever this code ends up, let's add a TODO comment here -- it will need to be updated if/when we support user-defined memory management styles

Reply:
Sounds good. I'll apply that to the logic further down that checks through a record's fields for nilable or non-nilable owned class types; maybe I can just put them both in their own function with one TODO pretty easily.